### ISSUE TITLE
[scan]: support `reinstall_app` option

### ISSUE THREAD

**Issue by dusek** [fastlane/fastlane#8038]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description

It would be great if scan supported a `reinstall_app` option, just like `snapshot` [does](https://github.com/fastlane/fastlane/blob/012e37642d49ad1612c6f7443e4d8a9712a143f6/snapshot/lib/snapshot/runner.rb#L223): before running the tests, it would first uninstall the app, only then install it and run the tests.

The difference is that during uninstall, (unlike during install which installs over the currently installed version of app and does not touch data), the data directories of the said app are removed too. E.g. documents, Core Data databases etc.

Our use case for this is Core Data. If we change the Core Data model in new version of app, then Core Data in the new version Core Data throws an exception if it starts with Core Data database from an older version of the model, and this fails tests immediately before they even started.

In our case the app is not yet released, so we prefer not to bother with any automatic or manual Core Data migrations as that would be loss of time. We will start adding data migrations only when it is first released.

Even then, having support for migrations would not solve this, because if you are testing multiple versions in parallel in CI, one with older version of model, one with new version, you can end up with situation when older model tries to load newer data.

So the only option to solve this seems to support reinstalling of the app.

There is an action to reset simulators, but that is too drastic and time consuming where uninstalling the specific app in question would suffice.

**Comment by oliverfoggin** [comment]
The best approach for testing with core data is to create a test core data stack that uses an in memory store.

That way the disk is never written to and so the conflicts you mention cannot happen.

This also has the side effect of speeding up the read and write of core data. It also means you will not be affecting the working "database" that you are using when running the app.

There are several tutorials about how to set up the core data stack in this way. Each depends on various ways that you are doing things. I have used JSQCoreDataKit to do this in the past. It makes things very easy :)

**Comment by ohayon** [comment]
Thanks for raising this feature request @dusek! This is definitely something that I think is worthy of being in _scan_. If you would like to submit a PR we would be happy to review it! I am going to close this issue for now which definitely does not mean the request is going away. We will reopen if/when we get a chance to bite off this work and please feel free to continue discussion here as well! 🐙
